charon/pkgs/signature.py

```diff
@@ -0,0 +1,165 @@
+"""
+Copyright (C) 2022 Red Hat, Inc. (https://github.com/Commonjava/charon)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+         http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import os
+import subprocess
+import asyncio
+import logging
+import gnupg
+from typing import Awaitable, Callable, List, Tuple
+from charon.storage import S3Client
+
+logger = logging.getLogger(__name__)
+
+
+def generate_sign(
+    package_type: str,
+    artifact_path: List[str],
+    top_level: str,
+    prefix: str,
+    s3_client: S3Client,
+    bucket: str,
+    key_id: str = None,
+    key_file: str = None,
+    passphrase: str = None
+) -> Tuple[List[str], List[str]]:
+    """ This Python function generates a digital signature for a list of metadata files using
+    the GPG library for uploads to an Amazon S3 bucket.
+
+        * Does not regenerate the existing metadata files when existing
+        * Returning all failed to generate signature files due to exceptions
+        * key_id: A string representing the ID of the RSA key to use for signing,
+        GPG command line tool is required when this parameter is not None.
+        * key_file: A string representing the location of the private key file.
+        * passphrase: A string containing the passphrase for the RSA key for key_id or key_file.
+
+    It returns a tuple containing two lists: one with the successfully generated files
+    and another with the failed to generate files due to exceptions.
+    """
+
+    if key_file is not None:
+        gpg = gnupg.GPG(gnupghome='~/.gnupg')
```

Comment:
Not sure if this .gnupg file is using the ~/.gnupg as the only file location in user's environment. If this is confirmed by stakeholder, we can use this way. Otherwise we need a configuration to let user be able to customize it.